core/src/main/scala/kafka/utils/ShutdownableThread.scala

```diff
@@ -63,6 +65,8 @@ abstract class ShutdownableThread(val name: String, val isInterruptible: Boolean
         doWork()
       }
     } catch{
+      case fatalExitException: FatalExitException =>
```

Comment:
Thanks for testing the patch @resetius . I updated the patch by 

i) Making exception inherit from Error (instead of Exception): this does not address the current issue but clarifies that FatalExitError is supposed to be handled by the framework rather than the application logic

ii) Hacking the code paths from handleOffsetOutOfRange to let FatalExitError reach the highest level.

The code seems to be full of abuse of Throwable (which are normally not supposed to be caught in the middle of application). On the other hand there are legit cases that we do not want the server shutdown if a request hits a bug and that is why it catches the general Throwable.

I am actually not quite sure how we want to proceed from this point. The proper fix would be to review all the usages of Throwable in the code, make its abuse to a minimum, and lets FataExitError to pass through in the few remaining cases. On the other hand, we might want to proceed with a minimum yet quick patch that fixes the current deadlock issue, and leave a proper fix for a later patch.

Thoughts?